File: Medium/Trapping Rain Water/trapping-rain-water.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>

using namespace std;


// } Driver Code Ends
class Solution{

    // Function to find the trapped water between the blocks.
    public:
    long long trappingWater(int arr[], int n){
        // code here
        //using 2 pointers
        int l=0;
        int r=n-1;
        long long maxL=0;
        long long maxR=0;
        long long ans=0;
        while(l<=r)
        {
            if(maxL<=maxR)
            {
                ans+=max((long long)0,maxL-arr[l]);
                maxL=max(maxL,(long long)arr[l]);
                l++;
            }
            else
            {
                ans+=max((long long)0,maxR-arr[r]);
                maxR=max(maxR,(long long)arr[r]);
                r--;
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main(){
    
    int t;
    //testcases
    cin >> t;
    
    while(t--){
        int n;
        
        //size of array
        cin >> n;
        
        int a[n];
        
        //adding elements to the array
        for(int i =0;i<n;i++){
            cin >> a[i];            
        }
        Solution obj;
        //calling trappingWater() function
        cout << obj.trappingWater(a, n) << endl;
        
    }
    
    return 0;
}
// } Driver Code Ends

File: Medium/Trapping Rain Water/trapping-rain-water.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;
import java.lang.*;


class Array {

	public static void main (String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine().trim()); //Inputting the testcases
		while(t-->0){
		  
		    //size of array
		    int n = Integer.parseInt(br.readLine().trim());
		    int arr[] = new int[n];
		    String inputLine[] = br.readLine().trim().split(" ");
		    
		    //adding elements to the array
		    for(int i=0; i<n; i++){
		        arr[i] = Integer.parseInt(inputLine[i]);
		    }
		    
		    Solution obj = new Solution();
		    
		    //calling trappingWater() function
		    System.out.println(obj.trappingWater(arr, n));
		}
	}
}


// } Driver Code Ends


class Solution{
    
    // arr: input array
    // n: size of array
    // Function to find the trapped water between the blocks.
    static long trappingWater(int arr[], int n) { 
        // Your code here
        long sum = 0;
        int left = 0;
        int right = n-1;
        int maxLeft = arr[left];
        int maxRight = arr[right];
        while(left<right)
        {
            if(maxLeft<=maxRight)
            {
                sum += maxLeft - arr[left];
                left++;
                maxLeft = Math.max(maxLeft,arr[left]);
            }
            else
            {
                sum += maxRight - arr[right];
                right--;
                maxRight = Math.max(maxRight,arr[right]);
            }
            
        }
        return sum;
    } 
}


